### PR TITLE
GRID-430: High-performant grid renderers (parts 1 & 2)

### PR DESCRIPTION
**Pull request by joneit**
See Jira [GRID-430](https://appoji.jira.com/browse/GRID-430) for details.

### CODE REVIEW

**Comment by joneit** [comment]
Partial renderer is here! — beginning with commit eafcd4c... However, needs more work. Right now it is only caching computed value for comparison on next render. To properly do this, we need to also handle look and feel changes — such as selected, hovered, font, colors, as well as changes in column width due to updated font and value.

**Comment by joneit** [comment]
With 6db31a6 partial rendering now considers the font, style, color, and background color (including cell selection and hovering). The only consideration left to do is cell icons.

**Comment by joneit** [comment]
With commit 467a608 partial rendering now also considers the font, style, color, and background color (including cell selection and hovering). The only consideration left to do is cell icons.

File: src/behaviors/Behavior.js [hunk 1]
```diff
@@ -293,10 +293,7 @@ var Behavior = Base.extend('Behavior', {
      * @return {object} Newly created default empty tablestate.
      */
     getDefaultState: function() {
-        return Object.create(
-            this.grid._getProperties(),
-            { columnProperties: { value: [] } }
-        );
+        return this.grid._getProperties();
```
Comment (Dwaynekj) [inline src/behaviors/Behavior.js]: this is what you were referring to in my ticket?

File: src/behaviors/cellProperties.js [hunk 2]
```diff
@@ -55,15 +55,16 @@ var cell = {
...
      * @memberOf Column#
      */
     getCellOwnProperties: function(rowIndex, dataModel) {
         var rowData;
         return (
             this.index >= 0 && // no cell props on row handle cells
-            (rowData = (dataModel || this.dataModel).getRow(rowIndex)) && // no cell props on non-existant rows
-            rowData.__META && rowData.__META[this.name] // undefined if not previously created
+            (rowData = (dataModel || this.dataModel).getRow(rowIndex)) && // no cell props on non-existent rows
+            rowData.__META && rowData.__META[this.name] ||
```
Comment (Dwaynekj) [inline src/behaviors/cellProperties.js]: Can you explain this line?

File: src/lib/Canvas.js [hunk 3]
```diff
@@ -783,6 +781,24 @@ function getCachedContext(canvasElement, type) {
         values = Object.getPrototypeOf(values);
     };
 
+    var conditionalsStack = [];
+
+    gc.clipSave = function(conditional, x, y, width, height) {
+        conditionalsStack.push(conditional);
+        if (conditional) {
```
Comment (Dwaynekj) [inline src/lib/Canvas.js]: Whats the adv. of caching the clipping region?

File: src/renderer/bundle-columns.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+function bundleColumns() {
+    var bundle, columnBundles = [],
+        gridProps = this.grid.properties,
```
Comment (Dwaynekj) [inline src/renderer/bundle-columns.js]: ?

File: src/renderer/bundle-rows.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+'use strict';
+
+function bundleRows(resetCellEvents) {
```
Comment (Dwaynekj) [inline src/renderer/bundle-rows.js]: ?

**Pull request merged**